### ISSUE TITLE
Internal error when adding imports to certain standard modules

### ISSUE THREAD

**Issue by bradcray** [chapel-lang/chapel#15255]
It seems that we have a bug in `ensureRequiredStandardModulesAreParsed()` and related logic in the compiler which prevents certain imports from being added to the standard modules.  This seems to trigger when our built-in modules require the module but the user code does not.  To reproduce:

1) Add `import Sort.ShallowCopy;` to partition() in modules/packages/Sort.chpl at line 909:

```c
  proc partition(Data: [?Dom] ?eltType,
                 lo: int, pivIdx: int, hi: int,
                 comparator)
  {
    import Sort.ShallowCopy;
    // The following section categorizes array elements as follows:             
    //                                                                          
    //   |  =  |  <  |  ?  |  >  |  =   |                                       
    //    lo    a     b   c     d     hi                                        
```

2) Compile hello.chpl.  This should generate the following internal error due to parser.cpp:

```
internal error: It seems an internal module is using a mod.submod form [/parser.cpp:415]
```

https://github.com/chapel-lang/chapel/blob/bf856458b2b1eec29ec23373d579814df5f3bef8/compiler/parser/parser.cpp#L414-L416

3) Add `use Sort;` to hello.chpl.  This should make it work again.

**Comment by bradcray** [comment]
@lydia-duncan: This shows how to reproduce the case we were chatting about.

**Comment by lydia-duncan** [comment]
Noting here so I don't lose the thoughts:
The code in that function is intended to trigger the parsing of internal module files.  A nested module import is difficult to determine the source file of at parse time, due to not having resolved the call expression needed to indicate it.  However, we will need to know what file the indicated module lives in, in case of:
- the nested module being the only thing relied upon by another file
- finding the nested module to parse when we proceed with the "submodules across files" implementation
and potentially other situations

**Comment by bradcray** [comment]
This was resolved in PR #15312, but I never thought to close the issue because it wasn't the main focus of the PR.
